jax/_src/lax/lax.py

```diff
@@ -836,8 +838,10 @@ def argmax(operand: Array, axis: int,
   return argmax_p.bind(operand, axes=(axis,),
                        index_dtype=dtypes.canonicalize_dtype(index_dtype))
 
-def reduce(operands, init_values, computation: Callable,
-           dimensions: Sequence[int]):
+def reduce(operands: T,
```

Comment:
I don't think the type variable is quite accurate unfortunately. For example, it's legal to pass a scalar integer as an argument but get back a JAX array.

Reply:
Yeah, that's true.  It's also true for practically every Jax function that scalar can be silently promoted to shape `()` arrays.  I think it's still better to annotate such cases using a type variable since:
- in the case of scalars, most people will be using an annotation like [this one](https://github.com/NeilGirdhar/tjax/blob/master/tjax/_src/annotations.py#L35), and there won't be a type error, and
- in the case of dataclass pytrees, the type variable will correctly constrain the type.

What do you think?

Reply:
I think no type annotations are better than incorrect type annotations.

Reply:
I wouldn't call them "incorrect".  It's no more incorrect than annotating something as `int` when actually it should be annotated as `numbers.Integral`.  At the end of the day, the point of annotations is to help users find bugs.  Annotating as `int` everywhere catches plenty of bugs.

Similarly, I'm annotating my Jax arrays as `numpy.typing.NDArray[np.integer[Any]]` (etc.) since there is no corresponding Jax type yet.  But because there's a bijection between Jax array types and Numpy array types, this catches practically all of the typing errors.

I understand the desire to make everything perfect.  I'm also a perfectionist.

What if we can find a solution with overloads that makes scalar inputs produce `Any` outputs, would that be acceptable to you?

Reply:
Something like this:
```python
from typing import Any, TypeVar, overload, Union

T = TypeVar('T')
Scalar = Union[int, float]

@overload
def f(x: Scalar) -> Any:
    ...

@overload
def f(x: T) -> T:
    ...

def f(x: Any) -> Any:
    return x

x = 2
reveal_type(f(x))  # Any
reveal_type(f((x,)))  # Tuple[builtins.int]
```

Reply:
Wouldn't a simpler and more correct type look like this:
```
def f(x: Union[Array, Scalar]) -> Array:
```
?

Reply:
@hawkinsp Yes, you're right.  (I'm assuming that you're correcting my first overload.)

Reply:
No, I'm saying "you don't need the overloads at all: that type suffices".

Reply:
You do need the overload because you can have aggregate pytree types like a `flax.struct.dataclass`.

Reply:
This is what may be lost in this discussion.  All of the Jax code that I write extensively uses dataclasses as aggregate pytree types.  This makes code much easier to read than using, for example, tuples.  It would be nice to have type inference work for these types.  That means that if `x` is an instance of `SomeClass`, then `stop_gradient(x)` should also be `SomeClass`.

Reply:
I removed the type variable for this function since as Jake points out "having no type annotations would be better than having incorrect type annotations".